### ISSUE TITLE
Extending and contribution to viewhelper reference

### ISSUE THREAD

**Issue by githubrsys** [FluidTYPO3/fluidtypo3org#27]
# Summary

### at the moment
* it is quite hard for users to interpret the VH Reference while "reallife examples" are missing
* it it's quite unknown for users how to contribute to the VH Reference easily 

## Solution

Take the good approach from the templating manual where a button in the top right corner resides and go one step further. 
  
Imagine there is a button on each VH Reference page to add additional "real life snippets" as appended contribution to understand the usage of each indivdual VH, it would gain the documentation quite fast as it is a neat ux to contribute with no hassle.

**Do not misunderstand** The FE Rendering of the XSD should stay as it is. But additional infomation and snippets shall be appended to each VH reference page as a second layer.

## Thoughts 

* Perhaps a contribution manifest should guide the contributors how to write good examples
  * Tell people to divide their examples into multiple files (with fixed structure) within a gist
  * Summary limited 2 - 3 Sentences as mandatory
  * normal and inline syntax mandatory
  * additional section for extra stories about a certain solution ?
  * ???
* Optional automated live preview of snippet through https://fluidtypo3.org/library/try-fluid-now.html ?
* how to handle version differences ?
* Perhaps it's possible to add new library code examples directly out of each individual VH page and have a persistent  matching by adding an automated tag for each VH and version
* dev: i hate nothing more as broken exmaples :+1: 
   * If there are valid examples posted they could be parsed ... technology for this is already there with builder and "Try Fluid now" ... Why not use that?
* from gbod Add examples to the comments the viewHelpers. Maybe there is a short way to connect those with this feature https://fluidtypo3.org/library/try-fluid-now.html :8ball:  Sounds good. Should then also have the possiblity to cycle through different mixed value options.

## Results if adopted

* lowered "boarding hurdle" into fluidtypo3 VH Business
* gained inspiration on the usage of VH's in different situations
* higher acceptance from TYPOSCRIPT & TEMPLAVOILA evangelists
* The better the documentation, the more users will use it
* ideally not more administrative input by admin if concept is found for managing version differences
 
## IMHO

* Jquery went that successful because of it's easiness of usage and documentation. Not because of technical superiority. 

# Contribute to this

everybody is invited to contribute to this, to make the fluidtypo3 experience even more fun !!

**Comment by coding46** [comment]
@githubrsys What is wrong with the library? https://fluidtypo3.org/library/code-examples.html Collect examples is the very reason for this part of the site.

**Comment by githubrsys** [comment]
@coding46 nothing is wrong with library. It's great to show off bigger contexts. 

But if it's about the dedicated usage of a VH the best place is IMHO within the VH page with additional examples. There are plenty of examples for this concept.  One of the famous ones is the jquery api see -> http://api.jquery.com/hover/

IMHO there remains no question.

**Comment by githubrsys** [comment]
# Mockup
<img src="http://i.imgur.com/uM4yQiG.png"/>

**Comment by NamelessCoder** [comment]
TL;DR: There are a few problems that might be too big for this to be feasible, but there's also a way to improve the situation enough to make some of it possible - in smaller steps.

#### Manifest for writing examples (code library)

The problem with this is that the gists that are added are very often copy/pastes from a working setup. The users who submit them would not expect to have to rewrite them to a certain standard and include all necessary usage examples. So while it is a good idea to have guidelines, I don't agree that an actual manifest with hard requirements is a good idea. Being Gist-based it should exactly be that type of published copy/paste-like content.

FYI: I currently re-word contributions to the code library in order to make a proper title and summary. There are so few submissions that this is not a hassle to do. And although it has not been necessary yet - I will of course reject examples that aren't useful (and contact the author).

#### Optional automated live preview of snippet

Is not a good idea simply because so many ViewHelpers and arguments use data types that are intentionally prevented by the live Fluid example. The doodle plugin is intended as a way to try the syntax and learn the very basics - it was never intended to be able to render complex fluid template code.

If that were our goal, we could simply create a View that attempts to render the code blocks - but I would predict that almost every example would inevitably fail - simply because none of the code block examples for ViewHelpers *define any variables that the example depends on*.

Another problem that must be considered is the version requirement. In order to render such examples for an older version of VHS, we would have to have separate rendering slaves with each version and installed just to be able to render those. Currently, that entails some 30+ rendering slaves...

So this, to me, does not sound technically feasible. At least not with the current way the hundreds of ViewHelpers are documented, nor with any of the methods we are considering here.

#### How to handle version differences ?

This is already an issue. None of our examples say which version of the packages those examples are for - we simply assume they are always for the most recently released version unless noted otherwise. This still has not become an issue, but I expect to remove code library entries when they become incompatible with the last released version.

I would not require users to submit along with library examples, the min/max version of each extension involved. I would instead rely on users contacting the author of the gist to have him updated, or contact us to have it removed.

#### Automatic submissions of examples when viewing ViewHelper docs

Although on the face of it this does sound like a good idea there are multiple complications:

* The submitted example has to be stored either as new files or edited into existing files in each extension (the alternative is a database that would quickly become a monster, considering the many parameters for version compatibility etc). So these changes has to happen via a git commit, and we categorically deny creating and merging git commits from a web interface without having a valid user with his own GitHub credentials and with the commit being created as a pull request.
* If we had a user with his or her own credentials we could like to editing the source of the ViewHelper file itself, but the examples that go into ViewHelper class doc comments require a very particular style of formatting. To be perfectly honest: we would prefer that skilled developers write these simply because they have to be so consistently formatted; also considering that argument documentation is done in actual PHP code.
* If theoretically implemented, we would only be able to allow such contributions for our own extensions. Fluid itself uses a different contribution approach with entirely different requirements (among other things, the dreaded CLA). And we would only be able to accept such contributions for the development branch - which would mean we can't link to editing on GitHub from any other view than that of our development branches (or the link might point to an incorrect version of the file - or a missing file). This limits the usefulness of those links greatly.

A small note about existing an existing method: we *do* have a convention in place which lets EXT:schemaker render special markdown files to include even more details in the documentation. This is particularly useful to further document each argument - because of the limited descriptions that are possible in PHP. However: those argument descriptions pertain to each version of the extension and they also have to be maintained - separately.

#### Automatic adding of library examples extracted from Fluid ViewHelpers

This would cause an immediate spamming of the code examples with completely redundant examples, e.g. multiple examples per ViewHelper each needing a specific title in order to be useful to humans. We would also face extracting such examples into the library for each version - and this then requires that we tag our library code examples with a min/max version (see above). All of this is in addition to the obvious requirements; being able to synchronise examples when changes occur in the ViewHelper docs for example.

As such this entails a huge body of work for a feature which is predicted to cause a massive output with no logical categorisation - we would have an extension context, sure, but imagine that the code library examples listing was 2,000+ examples and selecting an extension context only limited that to around 500 examples.

In the case of usage examples that go into the code library we should aim for quality over quantity. Good examples involving multiple ViewHelpers cover many more angles than thousands of tiny examples which have no descriptive context.

## Conclusion

I like the idea of user contributions for examples - obviously. But in the context of ViewHelpers, these contribution easements have so many technical requirements (for us, in both creating and maintaining the solution) that this way, is not feasible.

Automatic processing of examples is not a bad idea in an of itself, but it entails a huge backend just to extract the information and maintain the library in the most basic version and entails an even more complex setup of multiple slaves if it were to also perform live renderings.

Finally, further complexity in the code library (i.e. extension versions used, ViewHelpers used, etc) would have to be added by user input fields because such information cannot be extracted from any existing context (Gist API) and automatic extraction requires the same rendering slave complexity as described above, in addition to having to somehow detect which version(s) of a package would be able to render the code.

Hopefully this conclusion and the remarks above make it clear *why* we insist this is not a feasible approach.

### Alternative suggestion

The ideal situation would be one where a handful of skilled users sit down and write consistent usage examples in the PHP class files, using the proper rules for formatting. We then render those as source examples (already supported) and would then be able to remove the automatically generated usage examples (which currently spam every possible argument with a dummy value) without losing anything. And at that point, an "edit file" link is a great idea, when viewing ViewHelper docs on the development branch. The existing documentation would serve as a great starting point for others to add identically formatted documentation.

The added benefit of a solution that uses the current methods of documenting is that the documentation gets included in the XSD file and can be rendered by any source - not just EXT:schemaker. And we like compatible solutions!

If someone wants to arrange a workshop for that I wouldn't mind using some of the FluidTYPO3 funds to sponsor.

**Comment by NamelessCoder** [comment]
Will close this issue since automating this carries with it too many technical implications requiring too much work for our team to handle. I "CC" @cedricziel, @coding46 and @xf- on this in the hope that it will be remembered and given a few hours of attention on one of the coming sprints!
